sympy/polys/polytools.py

```diff
@@ -3816,6 +3817,89 @@ def nth_power_roots_poly(f, n):
 
         return r.replace(t, x)
 
+    def which_roots(f, candidates, real=True):
+        """
+        Given a superset of roots of f, finds which ones are
+        roots of f. Note this won't work properly if candidates is
+        not a superset of the roots of f. The list returned has
+        unique elements, is at most the length of the number of
+        unique roots of f, and preserves the order of the original
+        list of candidates. The `real` argument only controls
+        the root counting: if `real=True`, the `.count_roots()` method
+        is used (only supported over ZZ, QQ, or AlgebraicField); if
+        `real=False`, the degree is used.
```

Comment:
A docstring should start with a single line summary:

https://docs.sympy.org/latest/contributing/docstring.html

I would write this as:
```python
    def which_roots(f, candidates, real=True):
        """
        Find roots of a square-free polynomial ``f`` from ``candidates``.

        Explanation
        ===========

        If ``f`` is a square-free polynomial and ``candidates`` is a superset
        of the roots of ``f``, then ``f.which_roots(candidates)`` returns a
        list containing exactly the set of roots of ``f``. The domain must be
        :ref:`ZZ`, :ref:`QQ`, :ref:`ZZ_I`, :ref:`QQ_I` or :ref:`QQ(a)` and
        ``f`` must be univariate and square-free.

        If ``real=True`` then ``candidates`` must be a superset of the real
        roots of ``f`` and ``f.which_roots(candidates)`` returns the set of
        real roots of ``f``. Otherwise if ``real=False`` then ``candidates``
        must be a superset of all complex roots of ``f`` and
        ``f.which_roots(candidates, real=False)`` returns exactly the set of
        all complex roots of ``f``.

        """
```
I still think though that it would be better not to have the `real_roots=True/False` keyword argument. Rather it should just be two different functions like `which_real_roots` and `which_all_roots` or something.

Reply:
Thanks for writing this! I've also separated it out into `which_real_roots` and `which_all_roots`.